### ISSUE TITLE
LM GLM GLMM GAM GAMM

### ISSUE THREAD

**Issue by XiangyunHuang** [XiangyunHuang/MASR#8]
1. Ben Bolker
   - Generalized Linear Models https://bbolker.github.io/stat4c03/
   - Generalized Linear/Mixed models https://bbolker.github.io/morelia_2018
   - Fitting mixed models with (temporal) correlations in R https://bbolker.github.io/mixedmodels-misc/notes/corr_braindump.html
1. Broadening Your Statistical Horizons: Generalized Linear Models and Multilevel Models https://bookdown.org/roback/bookdown-bysh/
1. Elements of Statistical Learning https://www.dataschool.io/15-hours-of-expert-machine-learning-videos/
1. Teaching material for Bayesian Statistics https://florianhartig.github.io/LearningBayes/
1. General Additive Models with Mixed-Effect Smooths https://github.com/noamross/mixed-effect-gams
1. 关于广义线性混合效应模型的高频问题 https://bbolker.github.io/mixedmodels-misc/glmmFAQ.html
1. An introduction to Linear, Generalized, and Mixed/Multilevel models with R https://github.com/Pakillo/LM-GLM-GLMM-intro
1. Estimating treatment effects and ICCs from (G)LMMs on the observed scale using Bayes http://rpsychologist.com/GLMM-part1-lognormal
1.  GAM GLMM 建模和数据分析的例子 https://rpubs.com/corey_sparks
1. Bayesian spatial and spatiotemporal GLMMs with possible extremes [glmmfields](https://github.com/seananderson/glmmfields)
1. Correlated log-normal chain-ladder model https://magesblog.com/post/correlated-lognormal-chain-ladder-model/
1. https://menugget.blogspot.jp/2016/02/visualizing-model-predictions-in-3d.html
1. An R package for large scale estimation with stochastic gradient descent https://github.com/airoldilab/sgd  Taking advantage of the bigmemory package, sgd also operates on data sets which are too large to fit in RAM as well as streaming data.

**Comment by XiangyunHuang** [comment]
lme4 等 R 包都通过命名空间导入的方式调用函数

**Comment by XiangyunHuang** [comment]
https://github.com/ourcodingclub/CC-Linear-mixed-models
Introduction to linear mixed models
https://ourcodingclub.github.io/2017/03/15/mixed-models.html

**Comment by XiangyunHuang** [comment]
Bayesian Deep Net GLM and GLMM
https://github.com/VBayesLab/deepGLMM https://arxiv.org/pdf/1805.10157.pdf

将深度前馈神经网络引入广义线性混合效应模型

Deep feedforward neural networks (DFNNs) are a powerful tool for functional approximation. We describe flexible versions of generalized linear and generalized linear mixed models incorporating basis functions formed by a DFNN. The consideration of neural networks with random effects is not widely used in the literature, perhaps because of the computational challenges of incorporating subject specific parameters into already complex models. Efficient computational methods for high-dimensional Bayesian inference are developed using Gaussian variational approximation, with a parsimonious but flexible factor parametrization of the covariance matrix. We implement natural gradient methods for the optimization, exploiting the factor structure of the variational covariance matrix in computation of the natural gradient. Our flexible DFNN models and Bayesian inference approach lead to a regression and classification method that has a high prediction accuracy, and is able to quantify the prediction uncertainty in a principled and convenient way. We also describe how to perform variable selection in our deep learning method. The proposed methods are illustrated in a wide range of simulated and real-data examples, and the results compare favourably to a state of the art flexible regression and classification method in the statistical literature, the Bayesian additive regression trees (BART) method

**Comment by XiangyunHuang** [comment]
1. 参数估计
  1. 最小二乘
	 1. 普通最小二乘 ols
	 1. 偏最小二乘 pls
	 1. 迭代加权最小二乘 rwls 广义最小二乘 gls
  1. 极大似然估计
	  - 极大似然估计 bbmle
	  - 最小二乘估计 `nlme::gls` lm
	  - 剖面似然 profile likelihood      
1. 多元统计分析
   - Principal Components Analysis 主成分分析 `prcomp` 或者 `princomp` 或者 `psych::principal`
    - Factor Analysis 因子分析 `psych::fa`
    - Cluster Analysis 聚类分析 
       - 层次聚类 `stats::hclust` 和 `cluster::agnes` 函数
       - 快速聚类 `flashClust::hclust` 和 `fastcluster::hclust`
    - Discriminant Analysis 判别分析
    - Correspondence Analysis 对应分析
    - Canonical Correlation Analysis 典则相关分析
    - Multidimensional Scaling 多维标度分析
    - Conjoint Measurement Analysis 联合测量分析
1. 附录
   1. 运筹优化 梯度下降 L-BFGS算法 单纯性算法 割平面算法 罚函数

多元分布、参数估计、假设检验和各类似然估计作为模型的内容，不单独作为章节来描述

## 线性模型 LM

- 奇异值和特征值分解 （线性模型） 
   - `stats::lm` `stats::glm` `stats::nlm` 
   - `MASS::rlm` `MASS::lqs` `nlme::gls` 
   - [svd](http://github.com/asl/svd) 
   - [RSpectra](https://github.com/yixuan/RSpectra) 
   - RcppEigen https://github.com/RcppCore/RcppEigen `fastLm`
     - column-pivoted QR 分解
     - unpivoted QR 分解
     - LLT Cholesky 分解
     - LDLT Cholesky 分解
     - Jacobi SVD 分解
     - eigenvalue-eigenvector decomposition of X'X 
   - [biglm](https://cran.r-project.org/web/packages/biglm/index.html)  内存外回归，借助数据库
     - 增量 QR 分解

LM  Linear Models  线性模型 

- 大数据集`biglm::biglm`  [bigFastlm](https://github.com/jaredhuling/bigFastlm)
- 稳健线性模型使用 M 估计`MASS::rlm` `robust` `robustlmm` `robustbase`
- 快速拟合`RcppArmadillo::fastLm`  `RcppEigen::fastLm` 与 `lm` 的比较


## 多元分布

mvtnorm

多元正态分布 `mvtnorm::pmvnorm` 概率分布 `dmvnorm` 密度分布 `qmvnorm` 分位数
多元 T 分布 `mvtnorm::pmvt` 概率分布 `dmvt` 密度分布 `qmvt` 分位数 
逆卡方分布 `geoR::dinvchisq` The (Scaled) Inverse Chi-Squared Distribution
混合正态分布

extraDistr 和 distr 统计分布

## 高级主题

- 高维线性模型
- 高维广义线性模型
- 高维稀疏模型
- 动态线性模型 DLM Dynamic Linear Models
- 动态广义线性模型 DGLM Dynamic Linear Models
- 线性时间序列模型 LTS linear Time Series
  - 一元时间序列分析
  - 多元时间序列分析
- 非线性时间序列模型 NLTS nonlinear Time Series
- Statistical Inference and Sure Independence Screening via Ball Statistics [Ball](https://github.com/Mamba413/Ball)
- SIS: Sure Independence Screening [SIS](https://CRAN.R-project.org/package=SIS)

## 变量选择

- 2017 Extended Comparisons of Best Subset Selection, Forward Stepwise Selection, and the Lasso 
  [arxiv](https://arxiv.org/abs/1707.08692) [github](https://github.com/ryantibs/best-subset/) Trevor Hastie, Robert Tibshirani, Ryan J. Tibshirani

## 统计模型

**AER** (*ivreg, tobit*), **afex** (*mixed*), **base** (*aov, aovlist, lm, glm*), **betareg** (*betareg*), **blme** (*blmer, bglmer*), **brms** (*brmsfit*), **crch**, **countreg** (*zerontrunc*), **coxme**, **estimatr** (*lm_robust, iv_robust*), **feisr** (*feis*), **gam** (*Gam*), **gamm4** , **gamlss**, **gbm**, **gee**, **geepack** (*geeglm*), **GLMMadaptive** (*MixMod*), **glmmTMB** (*glmmTMB*), **gmnl**, **lfe** (*felm*), **lme4** (*lmer, glmer, nlmer, glmer.nb*), **MASS** (*glmmPQL, polr*), **mgcv** (*gam, gamm*), **nnet** (*multinom*), **nlme** (*lme, gls*), **ordinal** (*clm, clm2, clmm*), **plm**, **pscl** (*zeroinf, hurdle*), **quantreg** (*rq, crq, rqss*), **robust** (*glmRob, lmRob*), **robustbase** (*glmrob, lmrob*), **robustlmm** (*rlmer*), **rstanarm** (*stanreg, stanmvreg*), **survey**, **survival** (*coxph, survreg*), **truncreg** (*truncreg*), **VGAM** (*vgam, vglm*)

**hglm** **glm2** **gllvm**

- cglm: Fits Conditional Generalized Linear Models 拟合条件广义线性模型
  Estimates the ratio of the regression coefficients and the dispersion parameter in conditional generalized linear models for clustered data.

- [cgam](https://mirrors.tuna.tsinghua.edu.cn/CRAN/web/packages/cgam/index.html)

- FENmlm: Fixed Effects Nonlinear Maximum Likelihood Models
  Efficient estimation of maximum likelihood models with multiple fixed-effects. Standard-errors can easily and flexibly be clustered and estimations exported [@Laurent_2018_Efficient].

- DHARMa: Diagnostics for Hierarchical Regession Models https://github.com/florianhartig/DHARMa

  The 'DHARMa' package uses a simulation-based approach to create readily interpretable scaled (quantile) residuals for fitted generalized linear mixed models. Currently supported are generalized linear mixed models from 'lme4' (classes 'lmerMod', 'glmerMod'), generalized additive models ('gam' from 'mgcv'), 'glm' (including 'negbin' from 'MASS', but excluding quasi-distributions) and 'lm' model classes. Alternatively, externally created simulations, e.g. posterior predictive simulations from Bayesian software such as 'JAGS', 'STAN', or 'BUGS' can be processed as well. The resulting residuals are standardized to values between 0 and 1 and can be interpreted as intuitively as residuals from a linear regression. The package also provides a number of plot and test functions for typical model misspecification problems, such as over/underdispersion, zero-inflation, and residual spatial and temporal autocorrelation.

- bernor: Logit-Normal generalized linear mixed models (GLMM) using ordinary, independent and identically distributed Monte Carlo.  http://www.stat.umn.edu/geyer/bernor/
  Bernoulli Regression with Normal Random Effects (R package, not on CRAN) https://github.com/cjgeyer/bernor

- H-likelihood [@Lee_2018_Generalized] Monte Carlo likelihood inference for missing data models [@Sung_2007_Monte] hglm
  Inference based on mixed-effect models, including generalized linear mixed models with spatial correlations and models with non-Gaussian random effects (e.g., Beta). Variation in residual variance (heteroscedasticity) can itself be represented by a generalized linear mixed model. Various approximations of likelihood or restricted likelihood are implemented, in particular h-likelihood (Lee and Nelder 2001 <doi:10.1093/biomet/88.4.987>) and Laplace approximation.

- qle: Simulation-Based Quasi-Likelihood Estimation

  A simulation-based quasi-likelihood method (Baaske, M., Ballani, F., v.d. Boogaart, K.G. (2014)) for parameter estimation of parametric statistical models for which closed-form representations of distributional characteristics are unavailable and can only be obtained by computationally intensive simulations of the model.

- EMJMCMC: Bayesian model configuration, selection and averaging in complex regression contexts 作者主页 https://www.mn.uio.no/math/english/people/aca/aliaksah/
  paper and code https://github.com/aliaksah/EMJMCMC2016

- BayesianTools 广义线性混合效应模型之贝叶斯模型选择 <https://github.com/florianhartig/BayesianTools>
  General-purpose MCMC and SMC samplers, as well as plot and diagnostic functions for Bayesian statistics, with a particular focus on calibrating complex system models. Implemented samplers include various Metropolis MCMC variants (including adaptive and/or delayed rejection MH), the T-walk, two differential evolution MCMCs, two DREAM MCMCs, and a sequential Monte Carlo (SMC) particle filter.

- gamm4: Generalized Additive Mixed Models using 'mgcv' and 'lme4'
  Estimate generalized additive mixed models via a version of function gamm() from 'mgcv', using 'lme4' for estimation.

- sandwich: Robust Covariance Matrix Estimators
  Model-robust standard error estimators for cross-sectional, time series, clustered, panel, and longitudinal data.

- spTimer: Spatio-Temporal Bayesian Modelling
  Fits, spatially predicts and temporally forecasts large amounts of space-time data using [1] Bayesian Gaussian Process (GP) Models, [2] Bayesian Auto-Regressive (AR) Models, and [3] Bayesian Gaussian Predictive Processes (GPP) based AR Models for spatio-temporal big-n problems. Bakar and Sahu (2015) <doi:10.18637/jss.v063.i15>.

- blmeco: Data Files and Functions Accompanying the Book "Bayesian Data Analysis in Ecology using R, BUGS and Stan"
  Data files and functions accompanying the book Korner-Nievergelt, Roth, von Felten, Guelat, Almasi, Korner-Nievergelt (2015) "Bayesian Data Analysis in Ecology using R, BUGS and Stan", Elsevier, New York.


## 高维统计

- [glmhdfe](https://github.com/julianhinz/R_glmhdfe) Fast Estimation of GLMs with High-Dimensional Fixed Effects

- [Hmsc](https://github.com/hmsc-r/HMSC) high-dimensional multivariate models: Hierarchical model of species communities

- [gamboostLSS](https://github.com/boost-R/gamboostLSS) Boosting models for fitting generalized additive models for location, shape and scale (GAMLSS) to potentially high dimensional data.

- [swne](https://github.com/yanwu2014/swne) Similarity Weighted Nonnegative Embedding (SWNE), a method for visualizing high dimensional datasets

- [datamicroarray](https://github.com/ramhiser/datamicroarray) collection of small-sample, high-dimensional microarray data sets to assess machine-learning algorithms and models.

- [qtl2](https://github.com/rqtl/qtl2) QTL analysis software for high-dimensional data and complex cross designs

- [alpaca](https://github.com/amrei-stammann/alpaca) Fast and Feasible Estimation of Generalized Linear Models with High-Dimensional k-way Fixed Effects

- [ncvreg](https://github.com/pbreheny/ncvreg) Regularization paths for SCAD- and MCP-penalized regression models

- [grpreg](https://github.com/pbreheny/grpreg) Regularization paths for regression models with grouped covariates

- [multiDA](https://github.com/sarahromanes/multiDA) High Dimensional Discriminant Analysis in R

- [High-Dimensional-Statistics](https://github.com/luca590/High-Dimensional-Statistics) PCA, Factor Analysis, CCA, Sparse Covariance Matrix Estimation, Imputation, Multiple Hypothesis Testing

- [hdrm](https://github.com/pbreheny/hdrm) High-Dimensional Regression Modeling: Methodology, Applications, and Software 书籍

  


## 机器学习

[machine learning in R: preprocessing, cross-validation, lasso, decision trees, random forest, xgboost, superlearner](https://github.com/dlab-berkeley/Machine-Learning-in-R)

Trevor Hastie - Gradient Boosting Machine Learning
https://youtu.be/wPqtzj5VZus?list=PLNtMya54qvOFQhSZ4IKKXRbMkyLMn0caa

Abandon Statistical Significance http://www.stat.columbia.edu/~gelman/research/published/abandon_final.pdf


## 统计应用

> 基于大公司研究院的应用性学术论文

1. 金融风控模型，大规模逻辑回归
1. 方差分析、A/B Test 和多重比较多用于互联网广告数据，Geo-level 实验设计

- 广告方面的，Geo-level 实验设计，本质是贝叶斯分层模型，也是贝叶斯广义线性混合效应模型 [Geo-level Bayesian Hierarchical Media Mix Modeling](https://ai.google/research/pubs/pub46000)

- FaST linear mixed models for genome-wide association studies 全基因组研究 [@FastLMM_2011_Lippert] 微软
  https://www.nature.com/articles/nmeth.1681 https://github.com/akzaidi/fastlmm

- 从应用回到技术和算法，先了解业务，然后建模

- [可重复数据分析流](https://github.com/karthik/rstudio2019) How To Make Your Data Analysis Notebooks More Reproducible 


### R 语言落地场景

- 美团 https://tech.meituan.com/2018/08/02/mt-r-practice.html
- 比较成熟的有京东、豆瓣、ebay、阿里
- 国外使用 R 环境的公司 https://github.com/ThinkR-open/companies-using-r
- 互联网公司技术架构，微信/淘宝/微博/腾讯/阿里/美团点评/百度/Google/Facebook/Amazon/eBay的架构 https://github.com/davideuler/architecture.of.internet-product

**Comment by XiangyunHuang** [comment]
VGAM VGAM: Vector Generalized Linear and Additive Models
https://www.stat.auckland.ac.nz/~yee/VGAM
https://CRAN.R-project.org/package=VGAM 

Actuarial functions and heavy tailed distributions for R
精算函数和重尾分布
https://github.com/vigou3/actuar
https://CRAN.R-project.org/package=actuar

**Comment by XiangyunHuang** [comment]
https://rpsychologist.com/r-guide-longitudinal-lme-lmer

Using R and lme/lmer to fit different two- and three-level longitudinal models

常见模型结构及其对应的 R 代码公式

考虑将 https://www.xiangyunhuang.com.cn/2019/05/normal-hierarchical-model/ 分层正态、曲线生长模型实现
